components/BottomNavigation/src/MDCBottomNavigationBar.m

```diff
@@ -0,0 +1,414 @@
+    }
+    @catch (NSException *exception) {
+      if (exception) {
+        // No need to do anything if there are no observers.
+      }
+    }
+  }
+}
+
+- (void)removeBottomNavigationitemViews {
+  for (MDCBottomNavigationItemView *itemView in self.itemViews) {
+    [itemView removeFromSuperview];
+  }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath
+                      ofObject:(id)object
+                        change:(NSDictionary<NSKeyValueChangeKey,id> *)change
+                       context:(void *)context {
+  if (!context) {
+    NSInteger selectedItemNum = 0;
+    for (NSUInteger i = 0; i < self.items.count; i++) {
+      UITabBarItem *item = self.items[i];
+      if (object == item) {
+        selectedItemNum = i;
+        break;
+      }
+    }
+    MDCBottomNavigationItemView *itemView = _itemViews[selectedItemNum];
+    if ([keyPath isEqualToString:kMDCBottomNavigationBarBadgeColorString]) {
+      itemView.badgeColor = change[kMDCBottomNavigationBarNewString];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationBarBadgeValueString]) {
+      itemView.badgeValue = change[kMDCBottomNavigationBarNewString];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationBarImageString]) {
+      itemView.image = change[kMDCBottomNavigationBarNewString];
+    } else if ([keyPath isEqualToString:kMDCBottomNavigationBarTitleString]) {
+      itemView.title = change[kMDCBottomNavigationBarNewString];
+    }
+  }
+}
+
+- (UIEdgeInsets)mdc_safeAreaInsets {
+  UIEdgeInsets insets = UIEdgeInsetsZero;
+#if defined(__IPHONE_11_0) && (__IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0)
+  if (@available(iOS 11.0, *)) {
+
+    // Accommodate insets for iPhone X.
+    insets = self.safeAreaInsets;
+  }
+#endif
+  return insets;
+}
+
+#pragma mark - Touch handlers
+
+- (void)didTouchDownButton:(UIButton *)button {
+  MDCBottomNavigationItemView *itemView = (MDCBottomNavigationItemView *)button.superview;
+  itemView.circleHighlightHidden = NO;
+}
+
+- (void)didTouchUpInsideButton:(UIButton *)button {
+  for (MDCBottomNavigationItemView *itemView in self.itemViews) {
+    if (itemView.button == button) {
+
+      // Newly selected item
+      [itemView setSelected:YES animated:YES];
+      itemView.circleHighlightHidden = YES;
+    } else {
+
+      // Deselect all other items
+      [itemView setSelected:NO animated:YES];
+    }
+  }
+}
+
+- (void)didTouchUpOutsideButton:(UIButton *)button {
+  MDCBottomNavigationItemView *itemView = (MDCBottomNavigationItemView *)button.superview;
+  itemView.circleHighlightHidden = YES;
+}
+
+#pragma mark - Setters
+
+- (void)setItems:(NSArray<UITabBarItem *> *)items {
+  if (_items == items) {
+    return;
+  }
+  NSAssert(items.count >= 3, @"Need to have at least 3 items in navigation bar.");
+  NSAssert(items.count <= 5, @"Navigation bar has a maximum of 5 items.");
+
+  _items = items;
+
+  [self removeBottomNavigationitemViews];
```

Comment:
This method is a single loop that is only called from here.  Consider simplifying by removing this method completely and just having the loop with a comment.

Reply:
Done.